### DISCUSSION TITLE
How to extract not original images but cropped images

### DISCUSSION THREAD

**Discussion by yuki-tashiro** [pymupdf/PyMuPDF#3969] [Q&A] [answered]
Hello,

Thank you for your great work on this project!

I would like to know how I can extract the cropped portion of images in a PDF, instead of the original, uncropped images.

Currently, I'm able to extract the original images using page.get_image_bbox(), but this doesn't account for images that are cropped on the page. What I need is to extract only the part of the image that is visible on the PDF page, excluding the portions that may have been cropped out.

Here’s my code along with the input PDF and the output images:

```
import fitz  # PyMuPDF
from PIL import Image
import io

def extract_visible_images_from_pdf(pdf_path, output_folder):

    pdf_document = fitz.open(pdf_path)

    # Process each page
    for page_num in range(len(pdf_document)):
        page = pdf_document.load_page(page_num)
        
        # Render the entire page as an image
        pix = page.get_pixmap()
        page_image = Image.frombytes("RGB", [pix.width, pix.height], pix.samples)

        image_list = page.get_images(full=True)
        print(f"image_list: {image_list}")

        # Get all images on the page
        if image_list:
            for img_index, img_info in enumerate(image_list):
                xref = img_info[0]
                base_image = pdf_document.extract_image(xref)
                image_stream = io.BytesIO(base_image["image"])
                img = Image.open(image_stream)

                # Extract the image bytes from the PDF
                image_rect = page.get_image_bbox(img_info)
                visible_image = page_image.crop((
                    int(image_rect.x0),
                    int(image_rect.y0),
                    int(image_rect.x1),
                    int(image_rect.y1)
                ))

                img_output_path = f"{output_folder}/page_{page_num + 1}_image_{img_index + 1}.png"
                
                visible_image.save(img_output_path)
                print(f"Saved cropped image: {img_output_path}")
        else:
            print(f"No images found on page {page_num + 1}")

    pdf_document.close()

# pdf_path = "/content/drive/MyDrive/MatsuoLab/LLaVA-Med-JP/tp240424-01d_02.pdf"
pdf_path = "/content/input.pdf"
output_folder = "/content/"
extract_visible_images_from_pdf(pdf_path, output_folder)

```

Here is my original pdf.

- [input.pdf](https://github.com/user-attachments/files/17443323/input.pdf)

Extrated images

- ![page_1_image_1](https://github.com/user-attachments/assets/f7ea9595-3651-4994-8746-56a9d535f40c)
- ![page_1_image_2](https://github.com/user-attachments/assets/46070ebd-f950-409a-8369-945c8a2a206f)

Ideal extracted images

- ![cropped_image](https://github.com/user-attachments/assets/41affd05-9733-442e-8743-be13b009bf03)

Could you please guide me on how to modify the code to achieve this?

Thank you for your assistance!

**Comment by JorjMcKie** [comment]
This is no issue, but a typical Discussions item ... transferring.

**Reply by yuki-tashiro** [reply]
Thank you for transferring to the correct place.

**Comment by JorjMcKie** [comment, marked as answer]
This is not easy to achieve, because there is no definition of what "cropped" means.
Like in your example: you did not **_crop_** anything - you just covered some of the image parts. There is no way of determining which parts have remained visible.
If you however **_know_** the window of visibility, the obviously most canonic way is to make a pixmap of that region and save it.

**Reply by dserv-nh** [reply]
Are you sure there is no other way?
With the attached PDF opened in Master PDF Editor, I can see the original region and remove the cropping

<details><summary>Screenshots</summary>
<p>

![image](https://github.com/user-attachments/assets/c1eb41e9-dd41-4564-b0aa-398dc1e0d9a3)
![image](https://github.com/user-attachments/assets/5e90cfe8-0221-4048-bd2a-71d1342cfb07)

</p>
</details>

So couldn't you somehow calculate the cropped region by (visually?) comparing it with and without the clipping path, and then remove everything from the bounding box what's not visible without the clipping path?
I hope this makes sense
